### PR TITLE
Refactor CRPS and pinball loss, add represented interval option

### PR DESCRIPTION
**Pull request by MentReeze**
image

current: CRPSMethod = 'uniform'
new default: CRPSMethod = 'interval'

### CODE REVIEW

File: packages/openstef-beam/src/openstef_beam/metrics/metrics_probabilistic.py [hunk 1]
```diff
@@ -16,18 +16,24 @@
...
+
 import numpy as np
 import numpy.typing as npt
 
-from openstef_core.exceptions import MissingExtraError
+from openstef_beam.metrics.metrics_deterministic import pinball_loss
+from openstef_beam.metrics.metrics_helpers import represented_interval_weights
 from openstef_core.types import Quantile
 
+type CRPSMethod = Literal["interval", "uniform"]
```
Comment (lschilders) [inline packages/openstef-beam/src/openstef_beam/metrics/metrics_probabilistic.py]: This is not necessarily CRPS-specific, so I'd rename this type to maybe `QuantileWeightMethod`/`QuantileWeightingMethod`

File: packages/openstef-beam/src/openstef_beam/metrics/metrics_helpers.py [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+# SPDX-License-Identifier: MPL-2.0
+
+"""Helper functions shared between probabilistic and deterministic metrics."""
+
+import numpy as np
+import numpy.typing as npt
+
+
+def represented_interval_weights(
+    quantiles: npt.NDArray[np.floating],
```
Comment (lschilders) [inline packages/openstef-beam/src/openstef_beam/metrics/metrics_helpers.py]: Although not fully consistent in the rest of `openstef-beam` yet, but using the `Quantile` type everywhere is a good convention to follow.

File: packages/openstef-beam/src/openstef_beam/metrics/metrics_deterministic.py [hunk 3]
```diff
@@ -601,16 +647,8 @@ def relative_pinball_loss(
...
-    errors = y_true - y_pred
-    pinball_losses = np.where(
-        errors >= 0,
-        quantile * errors,  # Under-prediction
-        (quantile - 1) * errors,  # Over-prediction
-    )
-
     # Calculate mean pinball loss (weighted if weights provided)
-    mean_pinball_loss = np.average(pinball_losses, weights=sample_weights)
+    mean_pinball_loss = pinball_loss(y_true, y_pred, quantile=quantile, sample_weights=sample_weights)
```
Comment (lschilders) [inline packages/openstef-beam/src/openstef_beam/metrics/metrics_deterministic.py]: This variable name can be confused with the method with the same name in `metrics_probabilistic.py`. Maybe rename to something like `pinball_loss_quantile` or `pinball_loss_value`?

File: packages/openstef-beam/src/openstef_beam/metrics/metrics_probabilistic.py [hunk 4]
```diff
@@ -65,16 +75,18 @@ def crps(
...
+        for i, quantile in enumerate(quantiles)
+    ])
 
-    return float(np.average(sr.crps_quantile(y_true, y_pred, quantiles), weights=sample_weights))
+    if method == "interval":
+        quantile_weights = represented_interval_weights(quantiles)
+    else:
+        quantile_weights = np.full(len(quantiles), 1.0 / len(quantiles))
+
+    return float(2.0 * (quantile_weights @ per_quantile_loss))
```
Comment (lschilders) [inline packages/openstef-beam/src/openstef_beam/metrics/metrics_probabilistic.py:96]: Can you explain the factor 2.0 for the user in the docstring?

**Review by lschilders — COMMENTED** [review]
Good implementation and nice cleanup of the metrics computation. Some suggestions for improvement.

**Review by lschilders — DISMISSED** [review]
LGTM!

**Review by lschilders — APPROVED** [review]
